docs/contributing/codebase-overview.md

```diff
@@ -0,0 +1,262 @@
+---
+id: codebase-overview
+title: Codebase Overview
+layout: contributing
+permalink: contributing/codebase-overview.html
+prev: how-to-contribute.html
+next: design-principles.html
+---
+
+This section will give you an overview of the React codebase organization, its conventions, and the implementation.
+
+If you want to [contribute to React](/react/contributing/how-to-contribute.html) we hope that this guide will help you feel more comfortable making changes.
+
+We don't necessarily recommend any of these conventions in React apps. Many of them exist for historical reasons and might change with time.
+
+### Custom Module System
+
+At Facebook, internally we use a custom module system called "Haste". It is similar to [CommonJS](https://nodejs.org/docs/latest/api/modules.html) and also uses `require()` but has a few important differences that often confuse outside contributors.
+
+In CommonJS, when you import a module, you need to specify its relative path:
+
+```js
+// Importing from the same folder:
+var setInnerHTML = require('./setInnerHTML');
+
+// Importing from a different folder:
+var setInnerHTML = require('../utils/setInnerHTML');
+
+// Importing from a deeply nested folder:
+var setInnerHTML = require('../client/utils/setInnerHTML');
+```
+
+However, with Haste **all filenames are globally unique.** In the React codebase, you can import any module from any other module by its name alone:
+
+```js
+var setInnerHTML = require('setInnerHTML');
+```
+
+Haste was originally developed for giant apps like Facebook. It's easy to move files to different folders and import them without worrying about relative paths. The fuzzy file search in any editor always takes you to the correct place thanks to globally unique names.
```

Comment:
Yea I learned that since I first wrote this (as a draft) but couldn't figure out how to explain it without sounding even more confusing. Left it as is because it applies in our context.